### PR TITLE
store/tikv: use rbtree for cache region search

### PR DESCRIPTION
**Pull request by disksing**
@coocood @shenli @siddontang

### CODE REVIEW

File: store/tikv/kv.go [hunk 1]
```diff
@@ -196,20 +196,20 @@ func (s *tikvStore) SendKVReq(req *pb.Request, regionID RegionVerID) (*pb.Respon
 		req.Context = region.GetContext()
 		resp, err := client.SendKVReq(req)
 		if err != nil {
-			log.Warnf("send tikv request error: %v, try next peer later", err)
+			log.Warnf("send tikv request error: %v, ctx: %v, try next peer later", err, req.Context.String())
```
Comment (siddontang) [inline store/tikv/kv.go]: you can use %s if you implement String

File: store/tikv/region_cache.go [hunk 2]
```diff
@@ -45,24 +48,36 @@ func (c *RegionCache) GetRegionByVerID(id RegionVerID) *Region {
...
 	}
 	r, err := c.loadRegion(key)
-	return r, errors.Trace(err)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.insertRegionToCache(r)
+	return c.getRegionFromCache(key), nil
```
Comment (siddontang) [inline store/tikv/region_cache.go]: why not returning r directly?

Comment (disksing) [inline store/tikv/region_cache.go]: @siddontang Other thread may have inserted a Region with the same ID before, we should always return the Region in cache. I'll add a note here.

Comment (siddontang) [inline store/tikv/region_cache.go]: You have locked before, other can insert?

File: store/tikv/region_cache.go [hunk 3]
```diff
@@ -74,76 +89,83 @@ func (c *RegionCache) NextPeer(id RegionVerID) {
...
+func (c *RegionCache) dropRegionFromCache(verID RegionVerID) {
+	r, ok := c.regions[verID]
+	if !ok {
+		return
+	}
+	c.sorted.Delete(newRBItem(r))
+	delete(c.regions, r.VerID())
+}
+
+// loadRegion loads region from pd client, and pick the first peer as leader.
```
Comment (shenli) [inline store/tikv/region_cache.go]: picks

File: store/tikv/region_cache.go [hunk 4]
```diff
@@ -74,76 +89,83 @@ func (c *RegionCache) NextPeer(id RegionVerID) {
...
 }
 
-// loadRegion get region from pd client, and pick the random peer as leader.
+func (c *RegionCache) insertRegionToCache(r *Region) {
+	if _, ok := c.regions[r.VerID()]; ok {
+		return
+	}
+	old := c.sorted.ReplaceOrInsert(newRBItem(r))
+	if old != nil {
+		delete(c.regions, old.(llrbItem).toRegion().VerID())
```
Comment (shenli) [inline store/tikv/region_cache.go]: Consider version to decide which one to be deleted.

Comment (disksing) [inline store/tikv/region_cache.go]: @shenli It is more simple if we are sure that a Region contains our search key is in the cache. Besides, it won't matter if we use an old Region replace a new one, TiKV will reject the requests anyway.

**Comment by coocood** [comment]
LGTM

**Comment by disksing** [comment]
@siddontang @coocood PTAL.

**Comment by coocood** [comment]
LGTM

**Comment by shenli** [comment]
@disksing Any update?

**Pull request merged**